File: Daily Resources/Day11/DSA - Tree Solutions/tree_distances_ii.py
import sys
sys.setrecursionlimit(2 * 10**5 + 10)

n = int(input())
tree = [[] for _ in range(n + 1)]
for _ in range(n - 1):
    a, b = map(int, input().split())
    tree[a].append(b)
    tree[b].append(a)

subsize = [1] * (n + 1)
res = [0] * (n + 1)

def dfs1(u, parent):
    for v in tree[u]:
        if v != parent:
            dfs1(v, u)
            subsize[u] += subsize[v]
            res[u] += res[v] + subsize[v]

def dfs2(u, parent):
    for v in tree[u]:
        if v != parent:
            res[v] = res[u] - subsize[v] + (n - subsize[v])
            dfs2(v, u)

dfs1(1, 0)
dfs2(1, 0)
print(' '.join(map(str, res[1:])))

File: Daily Resources/Day11/DSA - Tree Solutions/tree_distances_ii.cpp
#include <bits/stdc++.h>
using namespace std;

const int N = 2e5 + 5;
vector<int> tree[N];
long long int subsize[N], res[N], n;

void dfs1(int u, int parent) {
    subsize[u] = 1;
    for (int v : tree[u]) {
        if (v == parent) continue;
        dfs1(v, u);
        subsize[u] += subsize[v];
        res[u] += res[v] + subsize[v];
    }
}

void dfs2(int u, int parent) {
    for (int v : tree[u]) {
        if (v == parent) continue;
        res[v] = res[u] - subsize[v] + (n - subsize[v]);
        dfs2(v, u);
    }
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(0);
    
    cin >> n;
    for (int i = 0, a, b; i < n - 1; ++i) {
        cin >> a >> b;
        tree[a].push_back(b);
        tree[b].push_back(a);
    }

    dfs1(1, 0);
    dfs2(1, 0);
    
    for (int i = 1; i <= n; ++i)
        cout << res[i] << " ";
}
